### PR TITLE
ostree update on IP secondary (reworked)

### PR DESCRIPTION
**Pull request by mike-sul**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,11 @@ Our versioning scheme is `YEAR.N` where `N` is incremented whenever a new releas
 
 ## [??? (unreleased)]
 
+### Added
+
+- Ostree update on IP Secondaries: [PR](https://github.com/advancedtelematic/aktualizr/pull/1518)
+- Basic file update on IP Secondaries: [PR](https://github.com/advancedtelematic/aktualizr/pull/1518)
```
Comment (patrickvacek) [inline CHANGELOG.md]: I would argue that the OSTree item here should be under "Fixed", as it has been supported for a while. Maybe it's worth saying something like "Fixed several issues with OSTree updates on IP Secondaries".

File: src/aktualizr_secondary/aktualizr_secondary.cc [hunk 2]
```diff
@@ -1,73 +1,75 @@
...
+
+  if (hasPendingUpdate()) {
+    // TODO: refactor this to make it simpler as we don't need to persist/store
+    // an installation status of each ECU but store it just for a given secondary ECU
+    std::vector<Uptane::Target> installed_versions;
+    boost::optional<Uptane::Target> pending_target;
+    storage_->loadInstalledVersions(ecu_serial_.ToString(), nullptr, &pending_target);
+    data::InstallationResult install_res =
+        data::InstallationResult(data::ResultCode::Numeric::kUnknown, "Unknown installation error");
+    LOG_INFO << "There is a pending update, try to apply it, update hash: " << pending_target_.sha256Hash();
```
Comment (patrickvacek) [inline src/aktualizr_secondary/aktualizr_secondary.cc]: Might be worth making this just a bit clearer. "Try to apply it" sounds like the user should take action. Maybe rephrase to "Pending update found; attempting to apply it. Target hash: ".

File: src/aktualizr_secondary/aktualizr_secondary.h [hunk 3]
```diff
@@ -1,50 +1,62 @@
...
-  int32_t getRootVersionResp(bool director) const;
-  bool putRootResp(const std::string& root, bool director);
-  bool sendFirmwareResp(const std::shared_ptr<std::string>& firmware);
-
-  static void extractCredentialsArchive(const std::string& archive, std::string* ca, std::string* cert,
-                                        std::string* pkey, std::string* treehub_server);
+  using Ptr = std::shared_ptr<AktualizrSecondary>;
+
+ public:
+  // TODO: free AktualizrSecondary from dependencies as much as possible, e.g. bootloader
```
Comment (patrickvacek) [inline src/aktualizr_secondary/aktualizr_secondary.h]: It has now been freed of the bootloader, at least from the perspective of the constructor here. :)

File: src/aktualizr_secondary/update_agent_file.cc [hunk 4]
```diff
@@ -0,0 +1,62 @@
+#include "update_agent_file.h"
+#include "logging/logging.h"
+#include "uptane/manifest.h"
+
+// TODO: this is an initial version of a file update on Secondary
```
Comment (patrickvacek) [inline src/aktualizr_secondary/update_agent_file.cc]: Is this `TODO` still relevant?

File: src/libaktualizr/primary/sotauptaneclient.cc [hunk 5]
```diff
@@ -1162,43 +1211,40 @@ std::vector<result::Install::EcuReport> SotaUptaneClient::sendImagesToEcus(const
...
-      f.first.install_res = data::InstallationResult(data::ResultCode::Numeric::kOk, "");
-      storage->saveInstalledVersion(f.first.serial.ToString(), f.first.update, InstalledVersionUpdateMode::kCurrent);
-    } else {
-      f.first.install_res = data::InstallationResult(data::ResultCode::Numeric::kInstallFailed, "");
+    if (fut_result == data::ResultCode::Numeric::kOk || fut_result == data::ResultCode::Numeric::kNeedCompletion) {
+      f.first.update.setCorrelationId(director_repo.getCorrelationId());
+      storage->saveInstalledVersion(f.first.serial.ToString(), f.first.update,
+                                    fut_result == data::ResultCode::Numeric::kOk
+                                        ? InstalledVersionUpdateMode::kCurrent
+                                        : InstalledVersionUpdateMode::kPending);
```
Comment (patrickvacek) [inline src/libaktualizr/primary/sotauptaneclient.cc]: Not a big deal or blocker here, but I'd probably prefer to have the three-line logic for calculating the result on its own line outside of the function call.

File: src/libaktualizr/primary/sotauptaneclient.h [hunk 6]
```diff
@@ -171,6 +173,7 @@ class SotaUptaneClient {
   // ecu_serial => secondary*
   std::map<Uptane::EcuSerial, std::shared_ptr<Uptane::SecondaryInterface>> secondaries;
   std::mutex download_mutex;
+  mutable Uptane::EcuSerial primary_ecu_serial_;
```
Comment (patrickvacek) [inline src/libaktualizr/primary/sotauptaneclient.h]: Why does this need to be `mutable`?

File: src/libaktualizr/uptane/isotpsecondary.h [hunk 7]
```diff
@@ -10,17 +10,17 @@ class IsoTpSecondary : public SecondaryInterface {
...
+  int getRootVersion(bool director) const override;
   bool putRoot(const std::string& root, bool director) override;
-  bool sendFirmware(const std::shared_ptr<std::string>& data) override;
-  Json::Value getManifest() override;
+  bool sendFirmware(const std::string& data) override;
+  Uptane::Manifest getManifest() const override;
 
  private:
-  IsoTpSendRecv conn;
+  mutable IsoTpSendRecv conn;
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/isotpsecondary.h]: Why is this `mutable` too?

File: src/aktualizr_secondary/uptane_verification_test.cc [hunk 8]
```diff
@@ -1,413 +0,0 @@
...
-}
-#endif
-
-extern "C" OstreeDeployment* ostree_sysroot_get_booted_deployment(OstreeSysroot* ostree_sysroot) {
-  return OstreeSecondaryUptaneVerificationTest::curOstreeDeployment(ostree_sysroot);
-}
-
-extern "C" const char* ostree_deployment_get_csum(OstreeDeployment* ostree_deployment) {
-  return OstreeSecondaryUptaneVerificationTest::curOstreeRootfsRev(ostree_deployment);
-}
```
Comment (patrickvacek) [inline src/aktualizr_secondary/uptane_verification_test.cc]: What happened to this test suite? In the old version there were a bunch of non-OSTree-specific Uptane tests. Those still seem useful to me.

File: src/aktualizr_secondary/aktualizr_secondary.cc [hunk 9]
```diff
@@ -78,146 +80,61 @@ int32_t AktualizrSecondary::getRootVersionResp(bool director) const {
...
-bool AktualizrSecondary::sendFirmwareResp(const std::shared_ptr<std::string>& firmware) {
-  auto targetsForThisEcu = director_repo_.getTargets(getSerial(), getHardwareID());
+bool AktualizrSecondary::putMetadata(const Metadata& metadata) { return doFullVerification(metadata); }
 
-  if (targetsForThisEcu.size() != 1) {
-    LOG_ERROR << "Invalid number of targets (should be one): " << targetsForThisEcu.size();
+bool AktualizrSecondary::sendFirmware(const std::string& firmware) {
+  // TODO: how to handle the case when secondary is rebooted after metadata are received
+  if (!pending_target_.IsValid()) {
+    LOG_ERROR << "No any pending target to receive update data/image for";
```
Comment (patrickvacek) [inline src/aktualizr_secondary/aktualizr_secondary.cc]: This message is also a bit confusing. Maybe change to "Aborting image download; no valid target found."

File: CHANGELOG.md (hunk 1 above)
Comment (mike-sul) [reply CHANGELOG.md]: ok

File: src/aktualizr_secondary/aktualizr_secondary.cc (hunk 2 above)
Comment (mike-sul) [reply src/aktualizr_secondary/aktualizr_secondary.cc]: ok

File: src/aktualizr_secondary/aktualizr_secondary.h (hunk 3 above)
Comment (mike-sul) [reply src/aktualizr_secondary/aktualizr_secondary.h]: yes

File: src/aktualizr_secondary/update_agent_file.cc (hunk 4 above)
Comment (mike-sul) [reply src/aktualizr_secondary/update_agent_file.cc]: well, it depends on what quality we would like to have here. At least, I am planning to add more tests for this case (file update)

File: src/libaktualizr/primary/sotauptaneclient.cc (hunk 5 above)
Comment (mike-sul) [reply src/libaktualizr/primary/sotauptaneclient.cc]: ok

File: src/libaktualizr/primary/sotauptaneclient.h (hunk 6 above)
Comment (mike-sul) [reply src/libaktualizr/primary/sotauptaneclient.h]: does not have to be such anymore, leftover of one the previous versions.

File: src/libaktualizr/uptane/isotpsecondary.h (hunk 7 above)
Comment (mike-sul) [reply src/libaktualizr/uptane/isotpsecondary.h]: because most of methods in the secondary interface are 'const' now and isotpsecondary uses IsoTpSendRecv non-const methods from within these const methods implementation.

File: src/aktualizr_secondary/uptane_verification_test.cc (hunk 8 above)
Comment (mike-sul) [reply src/aktualizr_secondary/uptane_verification_test.cc]: yes, the total number of tests has actually increased, I just split the initial test file into two, one with all non-ostree tests and another with ostree specific.

File: src/aktualizr_secondary/aktualizr_secondary.cc (hunk 9 above)
Comment (mike-sul) [reply src/aktualizr_secondary/aktualizr_secondary.cc]: ok

File: src/aktualizr_secondary/update_agent_file.cc (hunk 4 above)
Comment (patrickvacek) [reply src/aktualizr_secondary/update_agent_file.cc]: Oh, no problem. It can stay! I was just checking.

**Comment by mike-sul** [comment]
@patrickvacek updated according to the comments

File: src/libaktualizr/primary/sotauptaneclient.h (hunk 6 above)
Comment (patrickvacek) [reply src/libaktualizr/primary/sotauptaneclient.h]: Can you still remove the `mutable` modifier, or did it turn out it was necessary?

Comment (mike-sul) [reply src/libaktualizr/primary/sotauptaneclient.h]: sure, just missed it in the first time.
